File: leetcode/no0070_Climbing_Stairs.js
// link: https://leetcode.com/problems/climbing-stairs/

/**
 * @param {number} n
 * @return {number}
 */
let historyMap = {};
var climbStairs = function (n) {

    console.log("## INPUT:", n);

    if (n == 1) {
        return 1;
    }

    if (n == 2) {
        return 2;
    }

    if (n <= 0) {
        return 0;
    }

    if (historyMap.hasOwnProperty(n)) {
        return historyMap[n];
    }

    let steps = climbStairs(n - 1) + climbStairs(n - 2);
    historyMap[n] = steps;
    console.log("--> steps", steps);
    return steps;

};

File: leetcode/no0070_Climbing_Stairs.go
package main

import "fmt"

func climbStairs(n int) int {

	stepHistory := map[int]int{}

	var doClimb func(steps int) int

	doClimb = func(steps int) int {

		if steps == 1 {
			return 1
		}

		if steps == 2 {
			return 2
		}

		if steps <= 0 {
			return 0
		}

		_, isExisted := stepHistory[steps]
		if isExisted {
			return stepHistory[steps]
		}

		total := doClimb(steps-1) + doClimb(steps-2)
		stepHistory[steps] = total
		return total
	}

	return doClimb(n)
}

func main() {

	input := 2
	fmt.Println("INPUT:", input)
	result := climbStairs(input)
	fmt.Println("OUTPUT:", result)
}
